README.md
<img src="https://upload.wikimedia.org/wikipedia/commons/thumb/b/bc/BRA_orthographic.svg/270px-BRA_orthographic.svg.png" align="right" />

# viacep ![Tests](https://github.com/vnbrs/viacep/workflows/Tests/badge.svg) [![GitHub license](https://img.shields.io/github/license/vnbrs/burocracia.cr.svg)](https://github.com/vnbrs/burocracia.cr/blob/master/LICENSE)
> Dependency-less zipcode lookup gem for Brazilian addresses with an easy-to-use API and useful features such as timeout.

## Installing

You can easily install viacep through [RubyGems](https://rubygems.org/gems/viacep/):

```shell
gem install viacep
=> Rock on! 🚀
```

or add the following to your Gemfile:

```shell
gem "viacep"
```

## Usage

```ruby
require 'viacep'

# with unformatted zipcodes... =)
ViaCep::Address.new('80210130')
=> #<ViaCep::Address:0x00007fe52a8a0568 @cep="80210-130", @address="Rua José Ananias Mauad", @neighborhood="Jardim Botânico", @city="Curitiba", @state="PR", @ibge="4106902", @gia="">

# with formatted zipcodes... =)
ViaCep::Address.new('13035-680')
=> #<ViaCep::Address:0x00007fe52a99e730 @cep="13035-680", @address="Avenida João Jorge", @neighborhood="Vila Industrial", @city="Campinas", @state="SP", @ibge="3509502", @gia="2446">

# or even an nonexistent zipcode... =(
ViaCep::Address.new('12345678')
=> exception thrown: ViaCep::ApiRequestError

# and specifying a timeout in seconds... =)
ViaCep::Address.new('80210130', timeout: 0.2)
=> exception thrown: Timeout::Error (execution expired)
```

## Docs
The documentation is available on [RubyDoc](https://www.rubydoc.info/gems/viacep). Starting on [`ViaCep::Address`](https://www.rubydoc.info/gems/viacep/ViaCep/Address) might be a good start.

## Contributing

1. Create a fork (https://github.com/vnbrs/viacep/fork)
2. Create a branch (git checkout -b my-new-feature)
3. Make a commit  (git commit -am 'Add some feature')
4. Push your code (git push origin my-new-feature)
5. Create a Pull Request
6. Thanks! 🤙🏼


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/), and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## Unreleased
### Added
N/A

### Changed
N/A

### Deprecated
N/A

### Removed
N/A

### Fixed
N/A

### Security
N/A

## [2.0.3] - 2019-07-20
### Added
- Create CHANGELOG.md file to track changes.

### Fixed
- Improve public API documentation


spec/spec_helper.rb
require_relative '../lib/viacep'
require_relative './mock/request_mocker'
require 'webmock/rspec'

RSpec.configure do |config|
  config.include ViaCep::Test::RequestMocker

  config.expect_with :rspec do |expectations|
    expectations.include_chain_clauses_in_custom_matcher_descriptions = true
    expectations.syntax = :expect
  end

  config.mock_with :rspec do |mocks|
    mocks.verify_partial_doubles = true
  end

  config.shared_context_metadata_behavior = :apply_to_host_groups
  config.order = :random

  config.before(:each) do
    WebMock.reset!
    WebMock.disable_net_connect!
  end
end


spec/address_spec.rb
require 'spec_helper'

RSpec.describe ViaCep::Address do
  describe '#initialize' do
    context 'when a valid CEP is passed' do
      it 'returns the Address' do
        mock_viacep_success

        address = ViaCep::Address.new('80210130')
        expect(address.address).not_to be_empty
      end
    end

    context 'when a invalid length CEP is passed' do
      it 'raises ArgumentError' do
        expect { ViaCep::Address.new('000000') }
          .to raise_error(ArgumentError)
      end
    end

    context 'when a unexistent CEP is passed' do
      it 'raises ViaCep::ApiRequestError' do
        mock_viacep_not_found
        expect { ViaCep::Address.new('00000000') }
          .to raise_error(ViaCep::ApiRequestError)
      end
    end

    context 'when a nil CEP is passed' do
      it 'raises ArgumentError' do
        expect { ViaCep::Address.new(nil) }
          .to raise_error(ArgumentError)
      end
    end

    context 'when a low timeout is specified' do
      it 'raises a Timeout::Error' do
        mock_viacep_slow_success(delay_in_seconds: 0.01)
        expect { ViaCep::Address.new('80210130', timeout: 0.001) }
          .to raise_error(Timeout::Error)
      end
    end

    context 'when a regular timeout is specified' do
      it 'returns the Address' do
        mock_viacep_success

        address = ViaCep::Address.new('80210130', timeout: 3.0)
        expect(address.address).not_to be_empty
      end
    end
  end
end


spec/service_spec.rb
require 'spec_helper'

RSpec.describe ViaCep::Service do
  it 'should have a base URL' do
    expect(ViaCep::Service::BASE_URL).not_to be_empty
  end

  describe '.fetch' do
    context 'when a small timeout is specified' do
      it 'raises Timeout::Error' do
        mock_viacep_slow_success(delay_in_seconds: 0.01)
        expect { ViaCep::Service.fetch('80210130', 0.001) }
          .to raise_error(Timeout::Error)
      end
    end

    context 'when a suficient timeout is specified' do
      it 'returns the response' do
        mock_viacep_success
        response = ViaCep::Service.fetch('80210130', 5.0)
        expect(response).not_to be_empty
      end
    end

    context 'when a invalid CEP is passed' do
      it 'raises ViaCep::ApiRequestError' do
        mock_viacep_invalid_format
        expect { ViaCep::Service.fetch('000000', nil) }
          .to raise_error(ViaCep::ApiRequestError)
      end
    end
  end
end


lib/viacep/service.rb
require 'net/http'
require 'uri'
require 'json'
require 'timeout'
require_relative 'exceptions'

module ViaCep
  class Service
    BASE_URL = 'https://viacep.com.br/ws'.freeze

    #
    # Fetches the ViaCEP API to request a CEP.
    #
    # @example Fetch a CEP with no timeout.
    #   ViaCep::Service.fetch('80210130', nil)
    #   #=> {"cep"=>"80210-130", "logradouro"=>"Rua José Ananias Mauad", "complemento"=>"", "bairro"=>"Jardim Botânico", "localidade"=>"Curitiba", "uf"=>"PR", "unidade"=>"", "ibge"=>"4106902", "gia"=>""}
    #
    # @example Fetch a CEP with a timeout of 30 seconds.
    #   ViaCep::Service.fetch('80210130', 30)
    #   #=> {"cep"=>"80210-130", "logradouro"=>"Rua José Ananias Mauad", "complemento"=>"", "bairro"=>"Jardim Botânico", "localidade"=>"Curitiba", "uf"=>"PR", "unidade"=>"", "ibge"=>"4106902", "gia"=>""}
    #
    # @param [String] cep The CEP to be fetched.
    # @param [Integer, nil] timeout The timeout in seconds for the request to be finished.
    #
    # @raise [ViaCep::ApiRequestError] This is raised when the external API is down or the CEP does not exist.
    # @raise [Timeout::Error] This is raised when the timeout argument is specified and the request timed out.
    #
    # @return [Hash]
    #
    def self.fetch(cep, timeout)
      Timeout.timeout(timeout) do
        uri = URI("#{BASE_URL}/#{cep}/json")
        request = Net::HTTP.get_response(uri)

        if request.code == '200'
          response = JSON.parse(request.body)

          if response["erro"]
            raise ApiRequestError, "The server responded with HTTP 200 could not process the request"
          end

          response
        else
          raise ApiRequestError, "The server responded with HTTP #{request.code}"
        end
      end
    end
  end
end


lib/viacep/address.rb
require_relative 'service'

module ViaCep
  class Address
    attr_reader :cep, :address, :neighborhood, :city, :state, :ibge, :gia

    #
    # Initializes an instance of ViaCep::Address and fetches the CEP using the external API.
    #
    # @example Fetch a CEP
    #   ViaCep::Address.new('80210130')
    #   #=> #<ViaCep::Address:0x00007fe52a8a0568 @cep="80210-130", @address="Rua José Ananias Mauad", @neighborhood="Jardim Botânico", @city="Curitiba", @state="PR", @ibge="4106902", @gia="">
    #
    # @example Fetch a CEP with a formatted string
    #   ViaCep::Address.new('80210-130')
    #   #=> #<ViaCep::Address:0x00007fe52a8a0568 @cep="80210-130", @address="Rua José Ananias Mauad", @neighborhood="Jardim Botânico", @city="Curitiba", @state="PR", @ibge="4106902", @gia="">
    #
    # @param [String] cep The CEP to be fetched.
    # @option options [Integer] :timeout The timeout in seconds to the request.
    #
    # @raise [ArgumentError] This is raised when CEP is nil.
    # @raise [ArgumentError] This is raised when CEP format is invalid.
    # @raise [ViaCep::ApiRequestError] This is raised when the external API is down or the CEP does not exist.
    # @raise [Timeout::Error] This is raised when the timeout argument is specified and the request timed out.
    #
    # @return [ViaCep::Address]
    #
    def initialize(cep, options = {})
      if cep.nil?
        raise ArgumentError, 'CEP cannot be nil'
      end

      cep = cep.to_s.delete('^0-9')

      if cep.length != 8
        raise ArgumentError, 'CEP must have 8 digits'
      end

      response = Service.fetch cep, options[:timeout]
      fill_from response
    end

    private

    def fill_from(response)
      @cep          = response['cep']
      @address      = response['logradouro']
      @neighborhood = response['bairro']
      @city         = response['localidade']
      @state        = response['uf']
      @ibge         = response['ibge']
      @gia          = response['gia']
    end
  end
end
